quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/cli/Picocli.java

```diff
@@ -225,6 +234,59 @@ private static boolean hasProviderChanges() {
         return false;
     }
 
+    /**
+     * validate the expected values of non-cli properties
+     *
+     * @param cliArgs
+     * @param abstractCommand
+     */
+    public static void validateNonCliConfig(List<String> cliArgs, AbstractCommand abstractCommand) {
+        IncludeOptions options = getIncludeOptions(cliArgs, abstractCommand, abstractCommand.getName());
+
+        if (!options.includeBuildTime && !options.includeRuntime) {
+            return;
+        }
+
+        for (OptionCategory category : abstractCommand.getOptionCategories()) {
+            List<PropertyMapper> mappers = new ArrayList<>();
+            Optional.ofNullable(PropertyMappers.getRuntimeMappers().get(category)).ifPresent(mappers::addAll);
+            Optional.ofNullable(PropertyMappers.getBuildTimeMappers().get(category)).ifPresent(mappers::addAll);
+            for (PropertyMapper mapper : mappers) {
+                // bypass the PropertyMappingInterceptor - the transformations may cause unexpected errors
+                String value = null;
+                ConfigSource configSource = null;
+                for (ConfigSource cs : getConfig().getConfigSources()) {
+                    if (cs.getOrdinal() < 300) {
+                        break; // don't consider anything below standard env properties
+                    }
+                    value = cs.getValue(mapper.getFrom());
+                    if (value != null) {
+                        configSource = cs;
+                        break;
+                    }
+                }
+
+                if (value == null) {
+                    continue;
+                }
+
+                if (mapper.isBuildTime() && !options.includeBuildTime) {
+                    // TODO for cli this is treated as unknown
+                    continue;
+                }
+                if (mapper.isRunTime() && !options.includeRuntime) {
+                    // TODO for cli this is treated as unknown
+                    continue;
+                }
+
+                if (!PropertyMapperParameterConsumer.isExpectedValue(mapper.getExpectedValues(), value)) {
+                    throw new NonCliPropertyException(PropertyMapperParameterConsumer.getErrorMessage(mapper.getFrom(),
+                            value, mapper.getExpectedValues(), mapper.getExpectedValues()) + ". From ConfigSource " + configSource.getName());
+                }
+            }
```

Comment:
I'm not 100 % happy that we need to access config sources directly but I guess there's no other way if we want to get "unprocessed" value skipping the whole interceptor chain.

Reply:
Right the PropertyMappingInterceptor introduces a lot of considerations - especially with the build command.  It seems best to side-step that.  Otherwise there are some corrections that could / should be made to the PropertyMapper transformation / validation logic which is assuming that a valid db type is being used, or that a value exists (quite a few places are doing optional.get(), but optional could be empty).